pkg/kubectl/rolling_updater.go

```diff
@@ -307,26 +305,36 @@ func (r *RollingUpdater) scaleDown(newRc, oldRc *api.ReplicationController, desi
 	if oldRc.Spec.Replicas == 0 {
 		return oldRc, nil
 	}
-	// Block until there are any pods ready.
-	_, newAvailable, err := r.waitForReadyPods(config.Interval, config.Timeout, oldRc, newRc)
+	// Get ready pods. We shouldn't block, otherwise in case both old and new
+	// pods are unavailable then the rolling update process blocks.
+	// Timeout-wise we are already covered by the progress check.
+	oldAvailable, newAvailable, err := r.getReadyPods(oldRc, newRc)
 	if err != nil {
 		return nil, err
 	}
-	// The old controller is considered as part of the total because we want to
+	// The old available pods are considered as part of the total because we want to
 	// maintain minimum availability even with a volatile old controller.
 	// Scale down as much as possible while maintaining minimum availability
-	decrement := oldRc.Spec.Replicas + newAvailable - minAvailable
+	decrement := oldAvailable + newAvailable - minAvailable
 	// The decrement normally shouldn't drop below 0 because the available count
 	// always starts below the old replica count, but the old replica count can
 	// decrement due to externalities like pods death in the replica set. This
 	// will be considered a transient condition; do nothing and try again later
 	// with new readiness values.
 	//
-	// If the most we can scale is 0, it means we can't scale down without
-	// violating the minimum. Do nothing and try again later when conditions may
-	// have changed.
+	// If the most we can scale is 0, it means either 1) we can't scale down without
+	// violating the minimum so do nothing and try again later when conditions may
+	// have changed or 2) we can scale down old unready pods since they should not
+	// affect minAvailable.
 	if decrement <= 0 {
-		return oldRc, nil
+		// See if there are any unhealthy old pods; we can scale those down without
+		// violating minAvailable. Note that at this point minAvailable is already
+		// violated due to a rollout that successfully spinned up pods which later
+		// became unavailable.
+		if oldRc.Spec.Replicas <= oldAvailable {
+			return oldRc, nil
+		}
+		decrement = oldRc.Spec.Replicas - oldAvailable
```

Comment:
> One problem with this is that a flap of the pods of the oldRc will cause the rollout to advance unpredictably.

Also note that if oldRc pods become ready we still want to decrement. And we will based on L318